File: algorithm/22. Generate Parentheses/22.py
class Solution(object):
    def generateParenthesis(self, n):
        """
        :type n: int
        :rtype: List[str]
        """
        result=[]
        self.Parenthesis(result,"",n,n)
        return result
    
    def Parenthesis(self,result,string,l,r):
        if l==0 and r==0:
            result.append(string)
        if(l>0):
            self.Parenthesis(result,string+"(",l-1,r)
        if(r>0)and l<r:
            self.Parenthesis(result,string+")",l,r-1)

File: algorithm/22. Generate Parentheses/22.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> result;
        generateParenthesis(result,"",n,n);
        return result;
    }
    void generateParenthesis(vector<string>  &v, string s, int l, int r)
    {
        if(l==0 and r==0)
        {
            v.push_back(s);
        }
        if(l>0)
           generateParenthesis(v,s+'(',l-1,r);
        if(l<r && r>0)
            generateParenthesis(v,s+')',l,r-1);
    }
};
